### PR TITLE
Fix #8564: Make concrete inline methods final

### LINKED ISSUE #8564: Should concrete inline methods be effectively final?
In #8543 we align static and dynamic semantics of inline methods and allow inline methods to implement abstract methods.

There is one exception where static and dynamic semantics still differ: inline methods can override other inline methods. Example:
```scala
class A:
  def f: Int
class B extends A:
  inline def f = 1
class C extends B:
  override inline def f = 2

val b: B = C()
b.f  // gives 1, since dispatched statically
val a: A = b
a.f  // gives 2, since dispatched dynamically
```
051039e tried to close this loophole by disallowing overriding of concrete inline methods. But it breaks ScalaTest. See the discussion in #8543 for why this is the case. 

So, should we disallow this and require code that uses the pattern to be refactored? It might be cleaner in the end but could be painful to get there.

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/ast/Desugar.scala [hunk 1]
```diff
@@ -251,6 +251,8 @@ object desugar {
       cpy.DefDef(meth)(name = methName, tparams = tparams1), epbuf.toList)
 
     if (meth1.mods.is(Inline))
+      if !meth1.rhs.isEmpty then
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/ast/Desugar.scala]: I think it's cleaner to change the definition of `effectivelyFinal` instead.

File: compiler/src/dotty/tools/dotc/typer/RefChecks.scala [hunk 2]
```diff
@@ -157,7 +157,8 @@ object RefChecks {
    *    1.8.3  M is of type ()S, O is of type []T and S <: T, or
    *    1.9.1  If M is erased, O is erased. If O is erased, M is erased or inline.
    *    1.9.2  If M or O are extension methods, they must both be extension methods.
-   *    1.10   If O is inline, M must be inline
+   *    1.10.1. If O is inline, M must be inline
+   *    1.10.2. If M is inline, M must not be deferred
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/typer/RefChecks.scala]: Deferred inlines should be alllowed!

File: compiler/src/dotty/tools/dotc/typer/RefChecks.scala [hunk 3]
```diff
@@ -397,8 +398,10 @@ object RefChecks {
         overrideError("is an extension method, cannot override a normal method")
       else if (other.isAllOf(ExtensionMethod) && !member.isAllOf(ExtensionMethod)) // (1.9.2)
         overrideError("is a normal method, cannot override an extension method")
-      else if other.isInlineMethod && !member.isInlineMethod then // (1.10)
+      else if other.isInlineMethod && !member.isInlineMethod then // (1.10.1)
         overrideError("is not inline, cannot override an inline method")
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/typer/RefChecks.scala]: Should be redundant, if O is effectively final.

File: compiler/src/dotty/tools/dotc/typer/RefChecks.scala (hunk 2 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/typer/RefChecks.scala]: This is not an abstract inline. It is for an inline abstract override .

```scala
class A:
  def a: Int = 3

class B extends A:
  override inline def a: Int
```

```scala
class A:
  inline def a: Int = 3

class B extends A:
  override inline def a: Int
```

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/typer/RefChecks.scala]: I removed it

**Pull request merged**